doc/design/prune.md

```diff
@@ -0,0 +1,110 @@
+# Prune
+
+## Motivation
+
+We want to support running inference, training and checkpointing in one `ProgramDesc`. We implement 
+`void Prune(const ProgramDesc* input, ProgramDesc* output)` function, which takes a `ProgramDesc`
```

Comment:
1, need `target` in the interface
2, if we will not change input, than we should use reference
```cpp
void Prune(
        const ProgramDesc& input, 
        const std::vector<OpDesc*>& targets, 
        ProgramDesc* output) 
{...}
```

Reply:
1. I added `is_target` field in `OpDesc`
2. Sure.